Fix results screen pushed after rewinding in-between push delay

frenzibyte opened the pull request:
Closes #8793 

Although the issue is talking about the results screen pushed with current score after rewinding, as I've proposed in the thread, it's kind of annoying to push the results screen after rewinding from completion (by accident for example), so I've decided to solve this by cancelling the push delegate instead which technically solves this.

## Remarks
 - I wasn't initially sure about this solution but given that the `Player` push the results screen when all hit objects have been judged, this will do for now.

Commit:
Replace all judged event logic with HasCompleted bindable

Commit:
Use HasCompleted in Player

Commit:
Fix results screen pushed after rewinding in-between push delay

Commit:
Add test cases to ensure no regression in "cancelling completion"

bdach reviewed (commented):
As a side note I got the replay to pause after rewinding upon completion a few times (seems to be timing-dependent). Haven't looked in detail as to why, it could be a separate thing.

iiSaLMaN commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -412,12 +412,22 @@ public void Restart()
 
+            // cancel push delegate in case judges reverted
+            // after delegate may have already been scheduled.
+            if (!completionState.NewValue)
```
using `OldValue` in bool events is redundant here as I haven't used `BindValueChanged(..., true)` which makes `NewValue always the opposite of `OldValue`, therefore not adding anything.

btw the operations done inside this block is required to make the cancellation correct (hopefully), as they all reset everything that was set when a completion occurs:
 - `completionProgressDelegate?.Cancel()` – cancel the completion delegate to prevent from pushing. the delegate may already be null, so add a null-conditional, due to `allowResults` being false, thus not creating the delegate.
 - `completionProgressDelegate = null` – reset the delegate back to null to assure the `Player` that there isn't a push delegate scheduled.
- `ValidForResume = true;` – Reset back `ValidForResume` from being false to allow pausing back (though just remembered that rewinding comes only on replays which you can't pause, but to ensure that there are no traces of the player going through the completion process.)

bdach replied:
Fair point about the old value being redundant, although there was no need to explain the rest as it was pretty obvious even to me what it did.

iiSaLMaN replied:
ok great, *I was literally about to start commenting on each line in case something wasn't obvious*

peppy commented on osu.Game.Tests/Visual/Gameplay/TestSceneCompletionCancellation.cs:
```diff
@@ -0,0 +1,129 @@
+            AddStep("rewind to cancel", () => track.Seek(4000));
+            AddUntilStep("completion cleared by processor", () => !Player.ScoreProcessor.HasCompleted.Value);
+
+            AddWaitStep("wait", 5);
```
this seems prone to potential failure. can you make it an until step instead?

iiSaLMaN replied:
I don't think I can, this is a step that waits a couple of seconds before asserting to ensure that there was no attempt to push ranking screen, I will describe that in the step for clarity.

peppy replied:
In that case please make a const in player (`RESULTS_DISPLAY_DELAY`) and base the wait step duration on `RESULTS_DISPLAY_DELAY / TimePerAction * 2` or similar (`* 2` is to add some degree of safety), and add a comment.

iiSaLMaN replied:
Although with current values `(1000 / 200 * 2)`, `* 2` actually buffed up the wait count to `10`, I think ceiling is more appropriate here such that it doesn't fall one wait count behind.

peppy replied:
you may be able to reproduce an fail under visual tests due to the way scheduler works. make sure to test it thoroughly if you want to cut it that close.

Commit:
Centralize updating HasCompleted bindable logic

Commit:
Base wait steps duration on the delay used for results display

With `* 2` for safety of not potentially going to the next step and the delegate not executed yet.

Commit:
Use AddUntilStep whenever possible

Avoid redundant usage

Commit:
Merge remote-tracking branch 'upstream/master' into allow-cancelling-completion

peppy committed:
Merge branch 'master' into allow-cancelling-completion

peppy committed:
Refactor updateCompletionState implementation for legibility and code share

peppy committed:
Refactor tests
